ceno_zkvm/src/tables/ram/ram_impl.rs

```diff
@@ -425,6 +427,14 @@ impl<DVRAM: DynVolatileRamTable + Send + Sync + Clone> DynVolatileRamTableConfig
                     });
                 }
                 set_val!(row, self.final_cycle, rec.cycle);
+
+                let offset_addr = StructuralWitIn {
```

Comment:
this can move outside loop and create just once
```diff
diff --git a/ceno_zkvm/src/tables/ram/ram_impl.rs b/ceno_zkvm/src/tables/ram/ram_impl.rs
index 02cb6348..c84ac677 100644
--- a/ceno_zkvm/src/tables/ram/ram_impl.rs
+++ b/ceno_zkvm/src/tables/ram/ram_impl.rs
@@ -408,6 +408,11 @@ impl<DVRAM: DynVolatileRamTable + Send + Sync + Clone> DynVolatileRamTableConfig
             InstancePaddingStrategy::Custom(Arc::new(padding_fn)),
         );

+        let offset_addr = StructuralWitIn {
+            id: self.addr.id + (num_witness as u16),
+            ..self.addr
+        };
+
         final_table
             .par_iter_mut()
             .with_min_len(MIN_PAR_SIZE)
@@ -427,13 +432,6 @@ impl<DVRAM: DynVolatileRamTable + Send + Sync + Clone> DynVolatileRamTableConfig
                     });
                 }
                 set_val!(row, self.final_cycle, rec.cycle);
-
-                let offset_addr = StructuralWitIn {
-                    id: self.addr.id + (num_witness as u16),
-                    max_len: self.addr.max_len,
-                    offset: self.addr.offset,
-                    multi_factor: self.addr.multi_factor,
-                };
                 set_val!(row, offset_addr, rec.addr as u64);
             });
```